src/role_management.h

```diff
@@ -0,0 +1,49 @@
+#ifndef DQLITE_ROLE_MANAGEMENT_H
+#define DQLITE_ROLE_MANAGEMENT_H
+
+#include "server.h"
+
+struct all_node_info
+{
+	uint64_t id;
+	char *address;
+	int role;
+	bool online;
+	uint64_t failure_domain;
+	uint64_t weight;
+};
+
+/* Determine what roles changes should be made to the cluster, without
+ * side-effects. The given callback will be invoked for each computed change,
+ * with first argument the node whose role should be adjusted, second argument
+ * the node's new role, and third argument taken from the last argument of this
+ * function.
+ *
+ * The memory pointed to by @cluster is "borrowed" and not freed by this
+ * function, but it may be modified as part of this function's bookkeeping. */
+void RoleManagementComputeChanges(int voters,
+				  int standbys,
+				  struct all_node_info *cluster,
+				  unsigned n_cluster,
+				  dqlite_node_id my_id,
+				  void (*cb)(uint64_t, int, void *),
+				  void *arg);
+
+/* If necessary, try to assign new roles to nodes in the cluster to achieve
+ * the configured number of voters and standbys. Polling the cluster and
+ * assigning roles happens asynchronously. This can safely be called on any
+ * server, but does nothing if called on a server that is not the leader. */
+void RoleManagementAdjust(struct dqlite_node *d);
+
+/* Begin a graceful shutdown of this node. Leadership and the voter role will
+ * be transferred to other nodes if necessary, and then the callback will be
+ * invoked on the loop thread. The callback's second argument will be 0 if the
+ * handover succeeded and nonzero otherwise. */
+void RoleManagementHandover(struct dqlite_node *d,
+			    void (*cb)(struct dqlite_node *, int));
+
+/* Drain the queue of changes computed by RoleManagementAdjust. This should be
+ * done when the node is shutting down, to avoid a memory leak. */
+void RoleManagementDrain(struct dqlite_node *d);
+
```

Comment:
Nit: what about renaming:

- `roles_management.{c,h}` to  `roles.{c.h}`
- `RoleManagementAdjust()` to `RolesAdjust()`
- `RoleManagementHandover()` to `RolesHandover()`
- `RolesManagementDrain()` to `RolesCancelPendingChanges()`
- `RoleManagementComputeChanges` to `RolesComputeChanges()`
- `struct dqlite_node->changes` to `struct dqlite_node->roles_changes`

It's a bit more terse, and the "management" word is a bit redundant. We have many other things that are "management" (actually mostly everything :), but we don't tag them as such for sake of brevity.

Reply:
Right, what I meant is that it should be fairly obvious that `roles.c` is where role management code lives and that `RolesXXX` APIs are the ones dealing with role management.